### ISSUE TITLE
how to generate 2nd and 3rd embeddings?

### ISSUE THREAD

**Issue by Bio-bbhe** [mat10d/EvolvePro#16]
the way to get single mutant embeddings is very clear. my question is, for 2nd and 3rd embeddings, what are the parent sequences? E.g., for 2nd, are they all the possible single mutants relative to the best-n candidates from 1st round exp data? if so, what would be the number of n?

one more question, how does the performance of the ems model on multi-mutant?

**Comment by mat10d** [comment]
Well said @gky2023. Indeed a big challenge is how many combination mutations do you make available. For this reason, we restricted our multimutants to just the combination of single substitution mutants that had an activity score >1 (in an experimental setting, this usually means better that WT activity). This drastically reduces the number of multimutants to search over, but is a very simple way of approaching this, because indeed, two single mutants that have a decreased activity in combination may produce a higher activity mutant. So there is much work to be done in this area.

**Comment by ByteTora** [comment]
> Well said @gky2023. Indeed a big challenge is how many combination mutations do you make available. For this reason, we restricted our multimutants to just the combination of single substitution mutants that had an activity score >1 (in an experimental setting, this usually means better that WT activity). This drastically reduces the number of multimutants to search over, but is a very simple way of approaching this, because indeed, two single mutants that have a decreased activity in combination may produce a higher activity mutant. So there is much work to be done in this area.

I have a question: when generate a triple mutant, the round format is how? such as this code:
```python
round_name = 'Round5'
round_file_names_single = ['T7_pol_Round1.xlsx', 'T7_pol_Round2.xlsx', 'T7_pol_Round3.xlsx', 'T7_pol_Round4.xlsx']
round_file_names_multi = []
rename_WT = True

evolve_experimental_multi(
    protein_name,
    round_name,
    embeddings_base_path,
    [embeddings_file_name, embeddings_file_name_2nd, embeddings_file_name_3rd],
    round_base_path,
    round_file_names_single,
    round_file_names_multi,
    wt_fasta_path,
    rename_WT,
    number_of_variants,
    output_dir
)
round_name = 'Round6'
round_file_names_single = ['T7_pol_Round1.xlsx', 'T7_pol_Round2.xlsx', 'T7_pol_Round3.xlsx', 'T7_pol_Round4.xlsx']
round_file_names_multi = ['T7_pol_Round5.xlsx']
rename_WT = True

evolve_experimental_multi(
    protein_name,
    round_name,
    embeddings_base_path,
    [embeddings_file_name, embeddings_file_name_2nd, embeddings_file_name_3rd],
    round_base_path,
    round_file_names_single,
    round_file_names_multi,
    wt_fasta_path,
    rename_WT,
    number_of_variants,
    output_dir
)
```

in Round6, `T7_pol_Round5.xlsx`is also single mutants like this？
| Variant | Activity |
|---------|----------|
| 3F      | 0.91     |
| 4M      | 1.01     |
| 4Y      | 1.02     |
| 7A      | 0.82     |
| 10P     | 0.92     |
| 23K     | 0.99     |
| 23N     | 0.90     |

But from an experimental perspective, shouldn't the input data be like this(an example)?

| Variant | Activity |
|---------|----------|
| 4R_12C_31F      | 0.91     |
| 4P_12M_31A      | 1.01     |
| 4N_12P_31Q      | 1.02     |
| 4A_12N_31I      | 0.82     |

**Comment by zengxin95** [comment]
I have a question: when generate a triple mutant, the round format is how? like this?

<img width="172" alt="屏幕截图 2024-12-29 215302" src="https://github.com/user-attachments/assets/446a28dc-ecbb-451f-beb5-ef9cbe5532d2" />

**Comment by mat10d** [comment]
@zengxin95 @ByteTora yes, you are both correct. It is often easier, from an experimental perspective, to just have the single mutants listed as the position and the change (e.g. 3F). However, that becomes more complex when you have multi-mutants, because appending the original AA is difficult for a string of more than one substitution. So for multimutants, it is required that for each mutation, the original AA is listed (e.g. 4A_12N_31I).

**Comment by lzygitk7** [comment]
> [@zengxin95](https://github.com/zengxin95) [@ByteTora](https://github.com/ByteTora) yes, you are both correct. It is often easier, from an experimental perspective, to just have the single mutants listed as the position and the change (e.g. 3F). However, that becomes more complex when you have multi-mutants, because appending the original AA is difficult for a string of more than one substitution. So for multimutants, it is required that for each mutation, the original AA is listed (e.g. 4A_12N_31I).
---
I have a question about the format of 'beneficial_mutations.xlsx,' is the format like 

![Image](https://github.com/user-attachments/assets/d16e2e6b-7139-4475-8df3-1cbb3124d374)
where the first column is the top result from the last round of evolution? Does the second column represent the predicted score from the model?
Thank you in advance for your time.

**Comment by mat10d** [comment]
No, the second column is the actual scores produced by your wet-lab experiments.

**Comment by Hai-W-L** [comment]
@gky2023 Hi, I'm having the same issue as him @ByteTora . Below is my script code for performing multi-point mutations, but it's still generating single mutation results. How can I modify this?

from evolvepro.src.evolve import evolve_experimental, evolve_experimental_multi

protein_name = 'abdes3'
embeddings_base_path = '/home/spuser/EvolvePro/abdes/abdes3/plm'
embeddings_file_name = 'abdes3_esm2_t48_15B_UR50D.csv'
round_base_path = '/home/spuser/EvolvePro/abdes/abdes3/Round'
wt_fasta_path = "/home/spuser/EvolvePro/abdes/abdes3/raw/abdes3_WT.fasta"
number_of_variants = 20
output_dir = '/home/spuser/EvolvePro/abdes/abdes3/output'
rename_WT = False


# Multivariant
embeddings_file_name_2nd = 'dataset_2rd_esm2_t48_15B_UR50D.csv'


round_name = 'Round6'
round_file_names_single = ['Round1.xlsx','Round2.xlsx','Round3.xlsx','Round4.xlsx','Round5.xlsx']
round_file_names_multi = ['Round1_2rd.xlsx']
rename_WT = True

evolve_experimental_multi(
    protein_name,
    round_name,
    embeddings_base_path,
    [embeddings_file_name, embeddings_file_name_2nd],
    round_base_path,
    round_file_names_single,
    round_file_names_multi,
    wt_fasta_path,
    rename_WT,
    number_of_variants,
    output_dir
)

下面是输出：
Top 20 variants predicted by the model:
     variant     y_pred
1425    S75Y  12.854787
1424    S75W  12.842439
1411    S75F  12.832492
1408    S75C  12.829149
3431   S181M  12.825640
3439   S181Y  12.818802
3433   S181P  12.812038
3495   G184W  12.790752
3634   S192F  12.789862
1347    S71V  12.788770
3506   T185L  12.787059
562     T30M  12.786312
3422   S181C  12.783446
1338    S71I  12.782633
1422    S75T  12.782414
3602   T190M  12.782274
3647   S192W  12.779588
3436   S181T  12.778202
2699   S143A  12.778144
3648   S192Y  12.776172
